Add a "max_retries" field to work specs.

dmaze opened the pull request:
If a work spec data includes `"max_retries": 10`, then every time `Worker.RequestAttempts()` chooses that work spec and picks work units from it, count the number of attempts that already exist for that work unit, and if that count exceeds the `max_retries` limit, immediately fail the newly created attempt and look for more work.

Fixes #6.

Commit:
Add a "max_retries" field to work specs (without honoring it yet).

This is preserved in both the memory and PostgreSQL backends, and
appears in WorkSpecMeta.  It is documented everywhere, but as of this
change nothing actually causes failures as documented.

Commit:
Implement MaxRetries as described.

BurntSushi commented on postgres/attempt.go:
```diff
@@ -540,6 +562,59 @@ func (w *worker) requestAttemptsForSpec(req coordinate.AttemptRequest, spec *wor
+		// (It might be nice to do this in a batch, maybe moving
+		// this logic inside chooseAndMakeAttempts?)
+		for _, a := range moreAttempts {
+			// This should always be true but don't push it
```
Why not push it though? If `chooseAndMakeAttempts` should always return `[]*attempt`, why not panic when the invariant is broken? (It indicates a bug in the code?) Or do you imagine that `chooseAndMakeAttempts` could feasibly return other implementations of `coordinate.Attempt`?

dmaze replied:
Because Go won't cast a `[]*attempt` to a `[]coordinate.Attempt`.  😢 Given typical values it's not actually a lot of wasted cycles to do the conversion so I'll go in and do that.

BurntSushi replied:
Sorry, I don't think I was clear. What I meant was that the type convert here seems fine, but why not drop the `ok`? That is, explicitly panic if you don't get a `*attempt` since that indicates a bug?

dmaze replied:
I'd consider a `panic` here to be sloppy.  The change to return the concrete type I expect was pretty small, and I added it to this PR.

BurntSushi commented on postgres/attempt.go:
```diff
@@ -540,6 +562,59 @@ func (w *worker) requestAttemptsForSpec(req coordinate.AttemptRequest, spec *wor
+		// For each of the (new) attempts, count the number of
+		// existing attempts for the work unit and maybe fail it.
+		// (It might be nice to do this in a batch, maybe moving
+		// this logic inside chooseAndMakeAttempts?)
```
Do you have a feel for whether this will have a noticeable impact on performance or not? I don't quite have the cost model loaded into my head, but if a work spec has a non-zero max retry setting, then it looks like getting attempts has gone from "one query to get them all" to "one query to get them all, and then at least one additional extra query for each attempt." But I might be missing other steps that make this change more marginal than I think.

dmaze replied:
My immediate feel is that doing it this way involves somewhat more network I/O than doing a batch lookup but should be about the same net time (we want the database to do an index-only counting aggregation on an indexed value for each work unit).  It possibly increases the latency to get more work units, which could have a secondary effect of increasing the number of get-more-work conflicts.

My other intuition here is that the max-retries setting only really matters for computationally intensive Go work units, since the subprocess-based diffeo/python worker framework does a better job of forcibly killing and failing overtime work units (and the policy of "automatically retry at timeout" makes more sense there).

dmaze replied:
Also, the SQL query in `chooseAndMakeAttempts` is kind of intricate (it is a single query that chooses work units and creates a new attempt for each) and I'm a little afraid of adding another join to it.

BurntSushi replied:
> My other intuition here is that the max-retries setting only really matters for computationally intensive Go work units

Good point. Sold!

BurntSushi replied:
Oh, I see now where you wanted to go. Also a fine idea. :-)

BurntSushi reviewed (approved):
This LGTM. I left a couple of comments, but I'm happy with whatever response you might have!

Commit:
postgres: use internal types in internal attempt creation code.

Consistently use *attempt, not the coordinate.Attempt interface, inside
the guts of attempt creation.  This avoids a type cast (where we statically
know the type) when the max-retries logic needs to fail a newly created
attempt inside the current transaction, at the cost of copying the result
array at the outside of Worker.MakeAttempts.
